scala/data_structures/bloom_filter.scala

```diff
@@ -0,0 +1,55 @@
+package data_structures
+
+import scala.util.hashing.MurmurHash3
+import scala.util.Random
+
+case class BloomFilter[A](
+  nbOfItems: Int,
+  falsePositiveProbability: Double,
+  private val array: Array[Boolean],
+  private val hashFunctions: List[String => Int]
+) {
+
+  def +=(elem: A): BloomFilter[A] = {
+    val str = elem.toString
+    val hashes = this.hashFunctions.map(_(str))
+    val newArr = hashes.foldLeft(this.array)((acc, hash) => acc.updated(hash, true))
+    BloomFilter(nbOfItems, falsePositiveProbability, newArr, hashFunctions)
+  }
+
+  def ++=(elems: Seq[A]): BloomFilter[A] = elems.foldLeft(this)(_ += _)
+
+  def contains(elem: A): Boolean = {
+    val str = elem.toString
+    this.hashFunctions.forall(fn => this.array(fn(str)))
+  }
+}
+
+object BloomFilter {
+
+  def apply[A](nbOfItems: Int, falsePositiveProbability: Double): BloomFilter[A] = {
+    val arraySize = Math.ceil(Math.abs(nbOfItems * Math.log(falsePositiveProbability)) / Math.log(1 / Math.pow(Math.log(2), 2))).toInt
+    val array = Array.ofDim[Boolean](arraySize)
+
+    val nbOfHashFunctions = Math.round((arraySize / nbOfItems) * Math.log(2)).toInt
+    val hashFunctions: List[String => Int] =
+      (1 to nbOfHashFunctions).toList.map(_ => {
+        val seed = Random.nextInt
+        (str: String) => Math.abs(MurmurHash3.stringHash(str, seed)) % arraySize
+      })
+    BloomFilter(nbOfItems, falsePositiveProbability, array, hashFunctions)
+  }
+}
+
+object BloomFilterTest {
```

Comment:
tests could contain assertions right ?

Reply:
Yeah but those are not really tests, more like a quick demo. I didn't write tests for this actually (I know, TDD, shame on me :disappointed: ).
But actually the only thing that would make sense asserting upon is that the bloom filter *does not contain* certain elements, because it's _meant to_ return false-positives sometimes.
What would really be interesting though would be making statistics about the false positive rate depending on the arraySize and nbOfHashFunctions but some poeple have done this way before me so I'll trust them :sweat_smile:

Reply:
Yeah because it is a probabilistic data structure, it would probably be really complicated to test properly. Definitely interesting, though.

Reply:
👍